pkg/image/containersimageregistry/registry.go

```diff
@@ -0,0 +1,290 @@
+package containersimageregistry
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containers/common/pkg/auth"
+	"github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/pkg/docker/config"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	dockerconfig "github.com/docker/cli/cli/config"
+	"oras.land/oras-go/v2/content/oci"
+
+	orimage "github.com/operator-framework/operator-registry/pkg/image"
+)
+
+var _ orimage.Registry = (*Registry)(nil)
+
+type Registry struct {
+	sourceCtx *types.SystemContext
+	cache     *cacheConfig
+}
+
+var DefaultSystemContext = &types.SystemContext{OSChoice: "linux"}
+
+func New(sourceCtx *types.SystemContext, opts ...Option) (orimage.Registry, error) {
+	reg := &Registry{
+		sourceCtx: sourceCtx,
+	}
+
+	for _, opt := range opts {
+		if err := opt(reg); err != nil {
+			return nil, err
+		}
+	}
+
+	if reg.cache == nil {
+		var err error
+		reg.cache, err = getDefaultImageCache()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return reg, nil
+}
+
+func NewDefault() (orimage.Registry, error) {
+	return New(DefaultSystemContext)
+}
+
+type cacheConfig struct {
+	baseDir  string
+	preserve bool
+}
+
+func (c *cacheConfig) ociLayoutDir() string {
+	return filepath.Join(c.baseDir, "oci-layout")
+}
+func (c *cacheConfig) blobInfoCacheDir() string {
+	return filepath.Join(c.baseDir, "blob-info-cache")
+}
+
+func (c *cacheConfig) getSystemContext() *types.SystemContext {
+	return &types.SystemContext{
+		BlobInfoCacheDir: c.blobInfoCacheDir(),
+	}
+}
+
+type Option func(*Registry) error
+
+func getDefaultImageCache() (*cacheConfig, error) {
+	if dir := os.Getenv("OLM_CACHE_DIR"); dir != "" {
+		return newCacheConfig(filepath.Join(dir, "images"), true), nil
+	}
+	return getTemporaryImageCache()
+}
+
+func getTemporaryImageCache() (*cacheConfig, error) {
+	tmpDir, err := os.MkdirTemp("", "opm-containers-image-cache-")
+	if err != nil {
+		return nil, err
+	}
+	return newCacheConfig(tmpDir, false), nil
+}
+
+func newCacheConfig(dir string, preserve bool) *cacheConfig {
+	return &cacheConfig{
+		baseDir:  dir,
+		preserve: preserve,
+	}
+}
+
+func ForceTemporaryImageCache() Option {
+	return func(r *Registry) error {
+		var err error
+		r.cache, err = getTemporaryImageCache()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func InsecureSkipTLSVerify(insecureSkipTLSVerify bool) Option {
+	return func(r *Registry) error {
+		if r.sourceCtx == nil {
+			r.sourceCtx = &types.SystemContext{}
```

Comment:
We wouldn't use the `DefaultSystemContext` here?

Reply:
I don't think so, no. Because `InsecureSkipTLSVerify` will be called by `New`, which accepts an explicit `*types.SystemContext` from the caller. So if the user explicitly passes `nil` as the `*types.SystemContext`, we should not override that to the `DefaultSystemContext`.